File: sdk/python/examples/langchain/10_web_search_agent.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Web Search Agent — agent that simulates web search and page retrieval.

Demonstrates:
    - Simulated search result tools
    - Multi-step research: search → select → retrieve
    - Synthesizing information from multiple sources

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:6767/api
    - OPENAI_API_KEY for ChatOpenAI
"""

from langchain.agents import create_agent
from langchain_core.tools import tool
from langchain_openai import ChatOpenAI
from agentspan.agents import AgentRuntime


@tool
def web_search(query: str) -> str:
    """Search the web and return top 3 result summaries.

    Args:
        query: The search query string.
    """
    results = {
        "python history": [
            "Python was created by Guido van Rossum in 1989 and first released in 1991.",
            "Python 2.0 introduced list comprehensions and garbage collection (2000).",
            "Python 3.0, a major breaking change, was released in December 2008.",
        ],
        "machine learning": [
            "Machine learning is a branch of AI that enables systems to learn from data.",
            "Supervised learning uses labeled data; unsupervised learning finds hidden patterns.",
            "Deep learning uses neural networks with many layers for complex pattern recognition.",
        ],
        "climate change": [
            "Global average temperature has risen ~1.1°C above pre-industrial levels.",
            "CO2 levels reached 421 ppm in 2023, the highest in 3.6 million years.",
            "The Paris Agreement aims to limit warming to 1.5°C by reducing emissions.",
        ],
    }
    for key, items in results.items():
        if key in query.lower():
            return "\n".join(f"{i+1}. {item}" for i, item in enumerate(items))
    return f"Search results for '{query}': No cached results. (Demo mode — add more entries to results dict.)"


@tool
def get_page_content(url: str) -> str:
    """Retrieve the main content of a web page by URL.

    Args:
        url: The page URL to retrieve content from.
    """
    pages = {
        "python.org": "Python is a versatile, high-level programming language emphasizing readability. Used in web development, data science, AI, automation, and more.",
        "wikipedia.org/ml": "Machine learning (ML) is a field of AI research dedicated to developing systems that learn from data. Key techniques include regression, classification, clustering, and neural networks.",
    }
    for key, content in pages.items():
        if key in url.lower():
            return content
    return f"Content from {url}: (Demo mode — page content not cached.)"


@tool
def summarize_results(text: str) -> str:
    """Summarize a block of text into 2-3 key bullet points.

    Args:
        text: The text to summarize.
    """
    sentences = [s.strip() for s in text.split(".") if len(s.strip()) > 20]
    bullets = sentences[:3]
    return "\n".join(f"• {b}." for b in bullets) if bullets else "No content to summarize."


llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)
tools = [web_search, get_page_content, summarize_results]

graph = create_agent(
    llm,
    tools=tools,
    name="web_search_agent",
    system_prompt="You are a research assistant. Use search and retrieval tools to answer questions thoroughly.",
)

if __name__ == "__main__":
    with AgentRuntime() as runtime:
        result = runtime.run(
            graph,
            "Research the history of Python programming language and give me a brief summary.",
        )
        print(f"Status: {result.status}")
        result.print_result()

        # Production pattern:
        # 1. Deploy once during CI/CD:
        # runtime.deploy(graph)
        # CLI alternative:
        # agentspan deploy --package examples.langchain.10_web_search_agent
        #
        # 2. In a separate long-lived worker process:
        # runtime.serve(graph)

File: sdk/typescript/examples/langchain/10-web-search-agent.ts
/**
 * Web Search Agent -- agent that performs search and summarization with tools.
 *
 * Demonstrates:
 *   - Simulated web search tool returning structured results (mock data)
 *   - Page fetch tool returning mock content
 *   - ChatOpenAI reasoning about search results and summarizing
 *   - Running via AgentRuntime
 *
 * NOTE: This example uses mock search/fetch results for reproducibility.
 * For production, integrate Tavily, SerpAPI, or Brave Search.
 *
 * Requires: OPENAI_API_KEY environment variable
 */

import { ChatOpenAI } from '@langchain/openai';
import { DynamicStructuredTool } from '@langchain/core/tools';
import { HumanMessage, AIMessage, ToolMessage, SystemMessage } from '@langchain/core/messages';
import { RunnableLambda } from '@langchain/core/runnables';
import { z } from 'zod';
import { AgentRuntime } from '../../src/index.js';

// ── Mock search index (simulates web search API) ─────────

interface SearchResult {
  title: string;
  url: string;
  snippet: string;
}

const searchIndex: Record<string, SearchResult[]> = {
  langchain: [
    { title: 'LangChain Documentation', url: 'https://docs.langchain.com', snippet: 'LangChain is a framework for building applications with LLMs. It provides modules for chains, agents, memory, and retrieval.' },
    { title: 'LangChain GitHub', url: 'https://github.com/langchain-ai/langchain', snippet: 'Open-source Python and JavaScript library with 80k+ GitHub stars.' },
  ],
  langgraph: [
    { title: 'LangGraph Docs', url: 'https://langchain-ai.github.io/langgraph/', snippet: 'LangGraph is a library for building stateful multi-actor applications with LLMs, built on top of LangChain.' },
    { title: 'LangGraph Tutorial', url: 'https://blog.langchain.dev/langgraph/', snippet: 'LangGraph introduces graph-based orchestration of LLM workflows with support for cycles, branching, and persistence.' },
  ],
  python: [
    { title: 'Python.org', url: 'https://www.python.org', snippet: 'Python is a versatile, high-level programming language. The latest version is Python 3.13.' },
  ],
  openai: [
    { title: 'OpenAI API', url: 'https://platform.openai.com/docs', snippet: 'The OpenAI API provides access to GPT-4, DALL-E, Whisper, and Embeddings models via REST API.' },
  ],
};

const pageContent: Record<string, string> = {
  'docs.langchain.com': 'LangChain provides components including LLMs, PromptTemplates, Chains, Agents, and Memory. The LCEL allows composing these components with the | operator.',
  'langchain-ai.github.io/langgraph': 'LangGraph is built on top of LangChain and uses a graph-based approach where nodes are Python functions and edges define the flow between them. Supports cycles, persistence, and human-in-the-loop.',
  'python.org': 'Python 3.13 is the latest stable release. Key features include improved error messages and a free-threaded build option.',
  'platform.openai.com': "GPT-4o is OpenAI's most capable and efficient model. The API supports text, images, and function calling.",
};

// ── Tool definitions ─────────────────────────────────────

const webSearchTool = new DynamicStructuredTool({
  name: 'web_search',
  description: 'Search the web for information. Returns a list of search results with titles, URLs, and snippets.',
  schema: z.object({
    query: z.string().describe('The search query'),
  }),
  func: async ({ query }) => {
    const queryLower = query.toLowerCase();
    const results: SearchResult[] = [];
    for (const [keyword, entries] of Object.entries(searchIndex)) {
      if (queryLower.includes(keyword)) results.push(...entries);
    }
    if (results.length === 0) {
      return JSON.stringify([{
        title: `No results for '${query}'`,
        url: `https://search.example.com/?q=${encodeURIComponent(query)}`,
        snippet: `No cached results available for '${query}'.`,
      }]);
    }
    return JSON.stringify(results.slice(0, 3), null, 2);
  },
});

const fetchPageTool = new DynamicStructuredTool({
  name: 'fetch_page',
  description: 'Fetch and summarize the content of a web page given its URL.',
  schema: z.object({
    url: z.string().describe('The URL to fetch'),
  }),
  func: async ({ url }) => {
    for (const [key, content] of Object.entries(pageContent)) {
      if (url.includes(key)) return `Page content from ${url}:\n${content}`;
    }
    return `Page at ${url} returned general information. (Mock result)`;
  },
});

// ── Agent loop ───────────────────────────────────────────

const tools = [webSearchTool, fetchPageTool];
const toolMap = Object.fromEntries(tools.map((t) => [t.name, t]));

async function runSearchAgent(prompt: string): Promise<string> {
  const model = new ChatOpenAI({ modelName: 'gpt-4o-mini', temperature: 0 }).bindTools(tools);

  const messages: (SystemMessage | HumanMessage | AIMessage | ToolMessage)[] = [
    new SystemMessage(
      'You are a research assistant. Use web_search to find information, then optionally ' +
      'use fetch_page to get more details from specific URLs. Synthesize the results into ' +
      'a clear, well-organized summary with citations.'
    ),
    new HumanMessage(prompt),
  ];

  for (let i = 0; i < 6; i++) {
    const response = await model.invoke(messages);
    messages.push(response);

    const toolCalls = response.tool_calls ?? [];
    if (toolCalls.length === 0) {
      return typeof response.content === 'string'
        ? response.content
        : JSON.stringify(response.content);
    }

    for (const tc of toolCalls) {
      const tool = toolMap[tc.name];
      if (tool) {
        const result = await (tool as any).invoke(tc.args);
        messages.push(new ToolMessage({ content: String(result), tool_call_id: tc.id! }));
      }
    }
  }

  return 'Agent reached maximum iterations.';
}

// ── Wrap for Agentspan ───────────────────────────────────

const agentRunnable = new RunnableLambda({
  func: async (input: { input: string }) => {
    const output = await runSearchAgent(input.input);
    return { output };
  },
});

// Add agentspan metadata for extraction
(agentRunnable as any)._agentspan = {
  model: 'openai/gpt-4o-mini',
  tools,
  framework: 'langchain',
};

async function main() {
  const userPrompt = 'Search for information about LangGraph and summarize what you find.';

  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(agentRunnable, userPrompt);
    console.log('Status:', result.status);
    result.printResult();
  } finally {
    await runtime.shutdown();
  }
}

main().catch(console.error);
